### PR TITLE
copy rrs

### PR DESCRIPTION
**Pull request by miekg**
- Revert "make copies of RRs before returning them (#4409)"
- Document copying responses

### CODE REVIEW

**Comment by miekg** [comment]
cc @chrisohaver

**Comment by chrisohaver** [comment]
Is there a potential write/read race condition here when a plugin such as _file_ or _auto_ is updating an RR in the tree, at the same time the response writer is reading that RR?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] copy rrs (#44..." ]
>Is there a potential write/read race condition here when a plugin such as file
>or auto is updating an RR in the tree, at the same time the response writer is
>reading that RR?

likely yes... :-( so maybe not revert?

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] copy rrs (#44..." ]
>Is there a potential write/read race condition here when a plugin such as file
>or auto is updating an RR in the tree, at the same time the response writer is
>reading that RR?

although file doesn't update a single RR (no dynamic updates)

**Comment by chrisohaver** [comment]
> although file doesn't update a single RR (no dynamic updates)

Right - I suppose it replaces the entire tree, leaving the previous tree to be GC'ed. I see in file that only the tree/apex pointers are locked, not the contents of the trees. So, I think it's safe as is.  

Ultimately, putting the responsibility of fixing this on response writers seems better because it only copies when needed.

File: plugin.md [hunk 1]
```diff
@@ -131,6 +131,14 @@ The `fallthrough` directive should optionally accept a list of zones. Only queri
 in one of those zones should be allowed to fallthrough. See `plugin/pkg/fallthrough` for the
 implementation.
 
+## Mutating a Response
+
+If the plugin mutates a response it MUST make a copy of the entire response before doing so. A
```
Comment (chrisohaver) [inline plugin.md]:
```suggestion
Using a custom `ResponseWriter`, a plugin can mutate a response when another plugin further down the chain writes the response to the client.
If a plugin mutates a response it MUST make a copy of the entire response before doing so. A
```

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] copy rrs (#44..." ]
>    although file doesn't update a single RR (no dynamic updates)
>
>Right - I suppose it replaces the entire tree, leaving the previous tree to be
>GC'ed. I see in file that only the tree/apex pointers are locked, not the
>contents of the trees. So, I think it's safe as is.
>
>Ultimately, putting the responsibility of fixing this on response writers seems
>better because it only copies when needed.

Agreed

**Comment by miekg** [comment]
still needs fixed in rewrite

**Comment by chrisohaver** [comment]
> still needs fixed in rewrite

was expecting that to be added to this pr before it merged ...

**Comment by chrisohaver** [comment]
@miekg , Should we open an issue so we don't forget about it?  Or do you have a fix in the works already?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] copy rrs (#44..." ]
>    still needs fixed in rewrite
>
>was expecting that to be added to this pr before it merged ...

I suspect more plugins we need something, so didn't want to conflate and enlarge this
(which was essentally a revert only)

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] copy rrs (#44..." ]
>@miekg , Should we open an issue so we don't forget about it? Or do you have a
>fix in the works already?

feel free. I have no code laying around

**Pull request merged**